src/scene/shader-lib/program-library.js

```diff
@@ -133,11 +133,20 @@ class ProgramLibrary {
 
             // use shader pass name if known
             let passName = '';
+            let shaderPassInfo;
             if (options.pass !== undefined) {
-                const shaderPassInfo = ShaderPass.get(this._device).getByIndex(options.pass);
+                shaderPassInfo = ShaderPass.get(this._device).getByIndex(options.pass);
                 passName = `-${shaderPassInfo.name}`;
             }
 
+            // fire an event to allow the shader to be modified by the user. Note that any modifications are applied
+            // to all materials using the same generated shader, as the cache key is not modified.
+            this._device.fire('shader:generate', {
+                userMaterialId,
+                shaderPassInfo,
+                definition: generatedShaderDef
+            });
```

Comment:
nit: it would be great if this had a type so consumers could at least typecast the event `as pc.ShaderGenerateEvent` or similar